### DISCUSSION TITLE
Missing_ranges for a raw spss .sav data file

### DISCUSSION THREAD

**Discussion by steinunns** [Roche/pyreadstat#221] [Q&A] [answered]
I'm working with a raw .sav file.  I recoded all the missing values as 99. I am able to export the dataframe with the defined missing_ranges. However, I'm unable to read in the new .sav export unless I don't export the missing_ranges with the new .sav file. I'm guessing there is something wrong with how I'm creating the missing_ranges dictionary. 

I've tried 
`missing_ranges = {}
for item in column_list:
    missing_ranges[item] = [{'lo': 99, 'hi': 99}]`
output: 
{'birth': [{'lo': 99, 'hi': 99}],
 'grade': [{'lo': 99, 'hi': 99}]}

I have also tried having the dictionary look like:
{'birth': [99],
 'grade': [99]}

But I always get the same error when I try to read in the new .sav export:
![image](https://user-images.githubusercontent.com/49040186/226957133-e928f2f2-0956-4c9c-ae46-9a925e337cd7.png)

I've made sure there are no NaN values in the dataframe .

I've been able to export .sav files before with the missing ranges but that time the .sav file I was working with had already defined missing ranges and I just added a new item to the missing ranges dictionary after I added a new variable to the dataframe. 

Can someone help me?

**Comment by ofajardo** [comment, marked as answer]
very strange, there is something special in your data that causes the error I guess, because I have missing ranges in my tests and it works, see [here](https://github.com/Roche/pyreadstat/blob/master/tests/test_basic.py#L625). It looks very similar to what you are doing ... so I really don't know what is going on ... so some random question to check:

 have you tried for example writing missing ranges for only one variable and see if that works? 
Does it work if you use take away that meta_dict when writing?. Try to find what is the minimum of arguments to reproduce the issue.
Which version of pyreadstat are you using? the latest? in any case could you try another version to see if the problem is also there?
I guess it should not matter, I wrote 'hi' first and 'lo' second in my dictionary and you the opposite, but that shouldn't matter.

**Reply by ofajardo** [reply]
In case you are using the latest version of pyreadstat, it seems they have done some changes in the latest version of Readstat that may have something to do with this, it worth trying therefore the previous version of pyreadstat to see if the problem has been introduced in the most recent version. The opposite also holds true, maybe the issue is solved in the latest version, so you should try it if you haven't yet.

**Reply by steinunns** [reply]
That worked! Just had to update the pyreadstat package and restart my Jupyter notebook and everything works :)

**Reply by ofajardo** [reply]
wonderful!

**Comment by steinunns** [comment]
Now I noticed something strange, not sure if it happened in previous modules sine I'm pretty new working with .sav data. 

In my dataframe (same as above), the first 3 columns are descriptive (date, language and IDs) and are therefore not part of the metadata. The rest of the columns in the dataframe are of type integer and have variable value labels in the metadata except for 2 columns that hold answers for open ended questions and are of type object. These object columns are causing trouble. In the raw data these columns are the two last ones. When working with the data I created two new columns ('school', 'country'), 'school' of type object, and 'country' that's type int and added its information to the metadata dictionary and missing ranges. 

The trouble I'm getting is when I export the dataframe and try importing the new .sav file the new dataframe has 2 more columns than the dataframe that was exported and it has changed the names of my new columns and wiped out their data. It happens even though I export the dataframe without the variable_value_labels and missing ranges. I tried exporting it as a CSV and it worked fine so it was not the dataframe itself causing the problem. 

I tried rearranging the dataframe columns so that all non-integer columns appear first in the dataframe and exporting it that way along with the value labels and missing ranges. That works fine! I thought the variable_value_labels and missing_ranges uses the column names to match its properties and not the indices? Is this normal or something strange going on?

Screenshot of the end of the original dataframe
![image](https://user-images.githubusercontent.com/49040186/227542211-6a87052e-3c69-4326-96a7-8c0995819ec3.png)

Screenshot of the dataframe I get when I import the new .sav file
![image](https://user-images.githubusercontent.com/49040186/227541691-cc0c644d-8bda-4210-8a7f-7d9083672176.png)

The rearranged dataframe
![image](https://user-images.githubusercontent.com/49040186/227547117-3df3fc1c-1e18-4816-bc62-5ec3c3ccbb0b.png)

The imported .sav file from the rearranged dataframe
![image](https://user-images.githubusercontent.com/49040186/227546359-f6892f1a-ac1a-45fd-926f-e78e423ac12e.png)

**Reply by ofajardo** [reply]
very strange indeed, this one looks like a bug. If you isolate the error and do have a file you can share, you can create an issue to track the problem. It would be extremely useful if the data you share has the minimum number of variables and observations to reproduce the issue, i.e. the simpler the data, the better.

Do you have any very long string somewhere in your data, or do you have a variable with a very long name? Those seem to be causing problems some time. IF you have such things, try to remove them and see if that also fixes the issue.

**Reply by ofajardo** [reply]
oh, and could you please check if the issue is happening only in the last version of pyreadstat or also before?

**Comment by steinunns** [comment]
There was a very long string in one of string columns. When I removed that record I don't experience this error anymore. The longest string was 2800 characters, the second longest was around 500. Do you know if there is a max of "allowed" number of characters before it starts breaking the data?
I tried to see if the error also happens in previous versions so I  tried using pyreadstat version 1.1.8 and got the same error

**Reply by ofajardo** [reply]
OK, probably it is a variant of a known bug (see for [example](https://github.com/WizardMac/ReadStat/issues/260)), other people have reported somethig similar but only when opening the file in SPSS while it is fine in python ([here](https://github.com/Roche/pyreadstat/issues/119). 
Good thing is that it is tracked, bad thing is that it is not known when it is going to be solved. 
The limit I think is 756 C characters, but that may translate to lower number in python, since python uses unicode and some characters (non-ascii) are double.
